src/rust/engine/fs/src/snapshot.rs

```diff
@@ -341,6 +341,66 @@ impl Snapshot {
       .to_boxed()
   }
 
+  pub fn strip_prefix(
+    store: Store,
+    root_digest: Digest,
+    prefix: PathBuf,
+  ) -> impl Future<Item = Digest, Error = String> {
+    let store2 = store.clone();
+    Self::get_directory_or_err(store.clone(), root_digest)
+      .and_then(move |dir| {
+        futures::future::loop_fn(
+          (dir, PathBuf::new(), prefix),
+          move |(dir, already_stripped, prefix)| {
+            let mut components = prefix.components();
+            let component_to_strip = components.next();
+            if let Some(component_to_strip) = component_to_strip {
+              let remaining_prefix = components.collect();
+              let index = dir
+                .get_directories()
+                .binary_search_by(|node| {
```

Comment:
I believe that this should require that all entries in the snapshot are contained within the prefix, which would make this similar to applying [fast_relpath](https://github.com/pantsbuild/pants/blob/5eed2e7be5aa4672485f06e043bcec24bb9c668d/src/python/pants/util/dirutil.py#L35-L60) to all entries (ie fast-failing).